### PR TITLE
Make compilation of Linux source files conditional

### PR DESCRIPTION
**Pull request by pietern**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #206 Remove superfluous asserts
* #204 Fix C++14 extension warning when compiling for C++11
* #200 Use constexpr value for unspecified nbytes argument
* **#199 Make compilation of Linux source files conditional**
* #198 Export macro for every transport that gets compiled
* #197 Make aligned_allocator use posix_memalign

Differential Revision: [D16801686](https://our.internmc.facebook.com/intern/diff/D16801686)

### CODE REVIEW

File: gloo/common/CMakeLists.txt [hunk 1]
```diff
@@ -1,18 +1,25 @@
 set(GLOO_COMMON_SRCS
-  "${CMAKE_CURRENT_SOURCE_DIR}/linux.cc"
   "${CMAKE_CURRENT_SOURCE_DIR}/logging.cc"
   )
 
 set(GLOO_COMMON_HDRS
   "${CMAKE_CURRENT_SOURCE_DIR}/aligned_allocator.h"
```
Comment (mrshenli) [inline gloo/common/CMakeLists.txt]: (Also mentioned in #197), does `aligned_allocator.h` need to be moved as well, as `posix_memalign` is a linux specific call?

File: gloo/test/CMakeLists.txt [hunk 2]
```diff
@@ -5,17 +5,24 @@ set(GLOO_TEST_SRCS
...
   )
+set(GLOO_TEST_LIBRARIES)
+
+if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
+  list(APPEND GLOO_TEST_SRCS
+    "${CMAKE_CURRENT_SOURCE_DIR}/linux_test.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/multiproc_test.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport_test.cc"
+    )
+  list(APPEND GLOO_TEST_LIBRARIES rt)
```
Comment (mrshenli) [inline gloo/test/CMakeLists.txt]: what is `rt` here?

File: gloo/common/CMakeLists.txt (hunk 1 above)
Comment (pietern) [reply gloo/common/CMakeLists.txt]: macOS does have `posix_memalign`.

File: gloo/test/CMakeLists.txt (hunk 2 above)
Comment (pietern) [reply gloo/test/CMakeLists.txt]: `librt` contains functions including `clock_gettime`. This is used in `multiproc_test.cc`.

**Pull request closed without merging**